### PR TITLE
re: PeerManagerActor Make exceptions readable when TcpSocket::bind unwrap panics

### PR DESCRIPTION
**Pull request by pmnoxx**
While debugging test failures I noticed that `let listener = listener.unwrap();` can sometimes panic.
Let's improve error handling in `PeerManagerActor`.

### CODE REVIEW

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 1]
```diff
@@ -1516,7 +1517,7 @@ impl Actor for PeerManagerActor {
 
             ctx.spawn(TcpListener::bind(server_addr).into_actor(self).then(
                 move |listener, act, ctx| {
-                    let listener = listener.unwrap();
+                    let listener = listener.with_context(||format!("Failed to PeerManagerActor at {}", server_addr)).unwrap();
```
Comment (pmnoxx) [inline chain/network/src/peer_manager/peer_manager_actor.rs:1520]: @matklad Is there a way to format errors nicely, with similar syntax like with `tracing`

```
error!(message = "Failed to PeerManagerActor, server_addr)
```

**Comment by frol** [comment]
We might want to consider more [fine-grained approach to error handling when it comes to lib crates](https://nick.groenen.me/posts/rust-error-handling/), so `thiserror` might be a better choice than `anyhow`, yet this PR already brings net improvement, so we might want to merge it now, and come back to fine-grained error handling later.

**Comment by matklad** [comment]
If this is all we want to do in this PR, I suggest just 

```
listener.unwrap_or_else(|| panic!("Failed to PeerManagerActor at {}", server_addr));
```

there's no need for anyhow if what you want is just a nice panic message.

**Comment by pmnoxx** [comment]
> If this is all we want to do in this PR, I suggest just
> 
> ```rust
> listener.unwrap_or_else(|| panic!("Failed to PeerManagerActor at {}", server_addr));
> ```
> 
> there's no need for anyhow if what you want is just a nice panic message.

Actually, I want to add error propagation in the next PR. It's useful for unit tests.

**Comment by matklad** [comment]
> Actually, I want to add error propagation in the next PR. It's useful for unit tests.

I'd say that using `?` in tests is anti-pattern -- it's better if tests panic with an good error message, and you don't need anyhow for that, as it's possible to panic with format string. The reason why panic is better is that you immediately get a backtrace at the point of failure (anyhow supports backtraces only on nightly). You can also, at least in theory, configure the debugger to break on panics to speed up debugging (not sure if that actually works today -- I debug with prints :) ).

**Comment by pmnoxx** [comment]
> > Actually, I want to add error propagation in the next PR. It's useful for unit tests.
> 
> I'd say that using `?` in tests is anti-pattern -- it's better if tests panic with an good error message, and you don't need anyhow for that, as it's possible to panic with format string. The reason why panic is better is that you immediately get a backtrace at the point of failure (anyhow supports backtraces only on nightly). You can also, at least in theory, configure the debugger to break on panics to speed up debugging (not sure if that actually works today -- I debug with prints :) ).
 
I didn't mean using `?` in tests. Yes, that's not a good pattern I agree.

In tat message, I meant something different, let me explain.
We have been experiencing a lot of random test failures, some of them were caused by crashes happening inside our production code.
Often, I see crashes in out production code, while running tests, it takes a bit of time to figure out what the crash actually means. My plan is to propagate 
```
let listener = listener.with_context(||format!("Failed to PeerManagerActor at {}", server_addr)).unwrap();
```

**Comment by pmnoxx** [comment]
apply feedback

**Comment by matklad** [comment]
I think we need to rever lockfile changes and `use anyhow::Context`, but otherwise LGTM and is a big improvemen!

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 2]
```diff
@@ -30,6 +30,7 @@ use actix::{
     Actor, ActorFuture, Addr, Arbiter, AsyncContext, Context, ContextFutureSpawner, Handler,
     Recipient, Running, StreamHandler, WrapFuture,
 };
+use anyhow::Context as _Context;
```
Comment (matklad) [inline chain/network/src/peer_manager/peer_manager_actor.rs:33]: Note: you generally want `use anyhow::Context as _;`. That CLion bugs triggers in very rare circumstances, and I belive it was fixed after we reported it

**Review by matklad — APPROVED** [review]

**Review by bowenwang1996 — APPROVED** [review]

**Pull request merged**